federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+			// it always get the last service info from service store
+			s.enqueueService(obj)
+		}(obj, retryDelay)
+	} else if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to process service. Not retrying: %v", err))
+	}
+	return nil
+}
+
+// Returns an error if processing the service update failed, along with a time.Duration
+// indicating whether processing should be retried; zero means no-retry; otherwise
+// we should retry in that Duration.
+func (s *ServiceController) processServiceUpdate(cachedService *cachedService, service *v1.Service, key string) (error, time.Duration) {
+	// Ensure that no other goroutine will interfere with our processing of the
+	// service.
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+
+	// Update the cached service (used above for populating synthetic deletes)
+	// alway trust service, which is retrieve from serviceStore, which keeps the latest service info getting from apiserver
+	// if the same service is changed before this go routine finished, there will be another queue entry to handle that.
+	cachedService.lastState = service
+	err, retry := s.updateFederationService(key, service)
+	if err != nil {
+		message := "Error update service to all clusters"
+		if retry {
+			message += " (will retry): "
+		} else {
+			message += " (will not retry): "
+		}
+		message += err.Error()
+		s.eventRecorder.Event(service, api.EventTypeWarning, "UpdateServiceFail", message)
+		return err, cachedService.nextRetryDelay()
+	}
+	// Always update the cache upon success.
+	// NOTE: Since we update the cached service if and only if we successfully
+	// processed it, a cached service being nil implies that it hasn't yet
+	// been successfully processed.
+	//cachedService.appliedState = service
+	s.serviceCache.set(key, cachedService)
+	glog.V(4).Infof("successfully procceeded services %s", key)
+	cachedService.resetRetryDelay()
+	return nil, doNotRetry
+}
+
+// Returns an error if processing the service deletion failed, along with a time.Duration
+// indicating whether processing should be retried; zero means no-retry; otherwise
+// we should retry in that Duration.
+func (s *ServiceController) processServiceDeletion(key string) (error, time.Duration){
+	glog.V(2).Infof("Service %v is deleted, ensuring DNS Records is deleted", key)
+	cachedService, ok := s.serviceCache.get(key)
+	if !ok {
+		return fmt.Errorf("Service %s not in cache even though the watcher thought it was. Ignoring the deletion.", key), doNotRetry
+	}
+	service := cachedService.lastState
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	s.eventRecorder.Event(service, api.EventTypeNormal, "DeletingDNSRecord", "Deleting DNS Records")
+	// TODO should we delete dns info here or wait for endpoint changes? prefer here
+	// or we do nothing for service deletion
+	//err := s.dns.balancer.EnsureLoadBalancerDeleted(service)
+	s.deleteFederationService(service)
+	var err error
+	if err != nil {
```

Comment:
This code look messed-up?  err can only be nil here.
also deleteFederationService does not delete DNS records.  So the log messages and what's happeneing here are out of sync?